### PR TITLE
Feature/expandable design for proprietary data exchange

### LINKED ISSUE #1734: [SDL 0083] Expandable Design for Proprietary Data Exchange
Proposal: [Expandable Design for Proprietary Data Exchange](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0083-Expandable-design-for-proprietary-data-exchange.md)

> The SystemRequest/OnSystemRequest RPCs provide a mechanism for various components within the head unit to exchange data with app(s) written specifically to communicate with those components. With the growing popularity of SDL as a means of data exchange between head units and the outside world, there is a need to provide service to more such components. Today, the list of such components is limited to the 'RequestType' enum. Expanding this list everytime would not be an ideal solution. This document proposes a new mechanism for an expandable design in such scenarios by introducing a new string field called 'RequestSubType' to be used along with a new 'RequestType' enum called 'OEM_SPECIFIC'.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/249

Steering Committee Decision:

> When discussing this proposal, there was an ask for an example of the data that would be exchanged. Ford provided the example of an embedded navigation getting traffic information from a companion app. The Steering Committee then voted to accept this proposal.

### PR DESCRIPTION
**Pull request by AByzhynar**
Fixes #1734 

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
Unit tests are provided
ATF test scripts are provided

### Summary
- Made the following changes to Mobile API:
  Added new parameter "requestSubType" to SystemRequest and
  OnSystemRequest notification
  Added new RequestType - "OEM_SPECIFIC"

- Made the following changes to HMI API:
Added new parameter "requestSubType" to HMIApplication structure
and to OnAppPermissionChange notification
Added new RequestType - "OEM_SPECIFIC"

- Added handling of new parameters to policy regular
- Updated notifications triggering and their parameters handling
- Updated policy table validation and policy table update flow

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/appMain/sdl_preloaded_pt.json [hunk 1]
```diff
@@ -2344,7 +2344,9 @@
                 "steal_focus": false,
                 "priority": "NONE",
                 "default_hmi": "NONE",
-                "groups": ["Base-4"]
+                "groups": ["Base-4"],
+                "RequestType": [],
```
Comment (yang1070) [inline src/appMain/sdl_preloaded_pt.json]: what does  empty array  [] mean for "RequestType" and "RequestSubType"?
Does it mean allowing all request types/sub-types? or Does it mean allowing none of the request type/sub-types? For easy testing, probably the default value in preloaded_pt.json file, it shall allow all types.

Comment (yang1070) [reply src/appMain/sdl_preloaded_pt.json]: never mind, I got the answer from the comments in file policy_handler.cc, that [] mean allow everything.

Comment (AByzhynar) [reply src/appMain/sdl_preloaded_pt.json]: @yang1070 You are right. Sorry for my being late. You found the answer faster than I replied

**Comment by AByzhynar** [comment]
Rebuild required

File: src/components/policy/policy_regular/src/sql_pt_representation.cc [hunk 2]
```diff
@@ -744,6 +744,10 @@ bool SQLPTRepresentation::GatherApplicationPoliciesSection(
       return false;
     }
 
+    if (!GatherRequestSubType(app_id, &*params.RequestSubType)) {
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/policy/policy_regular/src/sql_pt_representation.cc]: @jacobkeeler No, we can't. RequestSubType is `Optional` - therefore has to be dereferenced (with `*`) , then we have to take an address of received object with `&`

File: src/components/policy/policy_external/src/policy_helper.cc [hunk 3]
```diff
@@ -322,9 +322,28 @@ bool CheckAppPolicy::operator()(const AppPoliciesValueType& app_policy) {
...
+    const bool is_request_type_changed = IsRequestTypeChanged(app_policy);
+    const bool is_request_subtype_changed = IsRequestSubTypeChanged(app_policy);
+    if (is_request_type_changed && is_request_subtype_changed) {
+      LOG4CXX_TRACE(
+          logger_,
+          "Both request types & subtypes were changed for application: "
+              << app_id);
+      SetPendingPermissions(app_policy,
+                            RESULT_REQUEST_TYPE_AND_SUBTYPE_CHANGED);
+      AddResult(app_id, RESULT_REQUEST_TYPE_AND_SUBTYPE_CHANGED);
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/policy/policy_external/src/policy_helper.cc]: @jacobkeeler We need it as SetPendingPermissions(...) has to set needed permissions changes. And there 3 available situations. So we need to separate results handling

File: src/components/policy/policy_regular/src/policy_helper.cc [hunk 4]
```diff
@@ -339,9 +339,22 @@ bool CheckAppPolicy::operator()(const AppPoliciesValueType& app_policy) {
...
-  if (!IsPredefinedApp(app_policy) && IsRequestTypeChanged(app_policy)) {
-    SetPendingPermissions(app_policy, RESULT_REQUEST_TYPE_CHANGED);
-    NotifySystem(app_policy);
+  if (!IsPredefinedApp(app_policy)) {
+    const bool is_request_type_changed = IsRequestTypeChanged(app_policy);
+    const bool is_request_subtype_changed = IsRequestSubTypeChanged(app_policy);
+
+    if (is_request_type_changed && is_request_subtype_changed) {
+      SetPendingPermissions(app_policy,
+                            RESULT_REQUEST_TYPE_AND_SUBTYPE_CHANGED);
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/policy/policy_regular/src/policy_helper.cc]: @jacobkeeler Unfortunately, no. We need it as SetPendingPermissions(...) has to set needed permissions changes. And there 3 available situations. So we need to separate results handling

File: src/components/application_manager/test/policy_handler_test.cc [hunk 5]
```diff
@@ -1377,12 +1377,32 @@ TEST_F(PolicyHandlerTest, IsRequestTypeAllowed) {
...
       mobile_apis::RequestType::eType::EMERGENCY;
+
+  EXPECT_CALL(*mock_policy_manager_, GetAppRequestTypesState(kPolicyAppId_))
+      .WillOnce(Return(policy::RequestType::State::AVAILABLE));
+
   EXPECT_CALL(*mock_policy_manager_, GetAppRequestTypes(kPolicyAppId_))
-      .WillOnce(Return(std::vector<std::string>()));
+      .WillOnce(Return(std::vector<std::string>({"HTTP"})));
   // Act
   policy_handler_.IsRequestTypeAllowed(kPolicyAppId_, type);
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/application_manager/test/policy_handler_test.cc]: @jacobkeeler Agree. It should.

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 6]
```diff
@@ -605,10 +606,8 @@ bool CommandRequestImpl::CheckAllowedParameters() {
   smart_objects::SmartMap::const_iterator iter_end = s_map.map_end();
 
   for (; iter != iter_end; ++iter) {
-    if (iter->second.asBool()) {
-      LOG4CXX_DEBUG(logger_, "Request's param: " << iter->first);
-      params.insert(iter->first);
-    }
+    LOG4CXX_DEBUG(logger_, "Request's param: " << iter->first);
+    params.insert(iter->first);
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/application_manager/src/commands/command_request_impl.cc]: @jacobkeeler Agree. These changes needed to this proposal as SDL makes parameters allowance check  by policy here. And RequestType & Subtype were skipped due to checks provided only for bool types. This fix is needed as this absolutely blocks correct ability to test this feature. So it was fixed here due to its `blocking` status

File: src/components/policy/policy_regular/src/policy_helper.cc [hunk 7]
```diff
@@ -422,7 +459,19 @@ void policy::CheckAppPolicy::SetPendingPermissions(
               EnumToJsonString(*it_request_type));
         }
       }
-
+      break;
+    case RESULT_REQUEST_SUBTYPE_CHANGED:
+      permissions_diff.priority.clear();
+      permissions_diff.requestSubTypeChanged = true;
+      {
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/policy/policy_regular/src/policy_helper.cc]: @jacobkeeler No. It's redundant in this case.

[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/policy/policy_regular/src/policy_helper.cc]: @jacobkeeler Fixed in 0637cd4cdc159ef60061094cf67ab9598a45d70b

File: src/components/application_manager/test/policy_handler_test.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/application_manager/test/policy_handler_test.cc]: @jacobkeeler Fixed in 0637cd4cdc159ef60061094cf67ab9598a45d70b

File: src/components/policy/policy_regular/src/policy_helper.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/policy/policy_regular/src/policy_helper.cc]: @jacobkeeler I got your idea. I will change it.

**Comment by AByzhynar** [comment]
Rebuild required

**Pull request merged**